rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,263 @@
+an Access Decision Function in some literature.
+
+- Policy Enforcement Point (PEP): A component responsible for enforcing access-control decisions. Sometimes referred to
+as an Access Enforcement Function in some literature.
+
+- Decision Request: A request sent to the PDP describing an action and requesting a decision on whether or not
+the action should be allowed, and any necessary information needed to correctly enforce limitations upon the action.
+
+- Permit: A single data structure encoding all necessary information for a PEP to correctly enforce a conditional/
+parameterized allow decision.
+
+### Overview
+
+We will rework teleport's access-control decision making to have a well-defined boundary between Policy Decision Point
+and Policy Enforcement Point. The goal of this rework will be to abstract as much complexity as possible into the PDP,
+while attempting to minimize the need to actually change how or where enforcement is done today. The PDP/PEP boundary
+will be representable as a gRPC API, but where possible we will continue to ensure that teleport does not make unnecessary
+network round trips.
+
+We will also establish a set of conventions intended to make the transition to using the new system as easy as possible,
+and to ensure that it is as difficult as possible to accidentally misuse it either by misunderstanding or oversight.
+
+### API
+
+The PDP/PEP boundary will be represented as a gRPC API, with methods for the different kinds of decisions. Because teleport
+as a whole is an integrated PDP/PEP, our decisions are often much more nuanced than a simple allow/deny. This is one of
+teleport's greatest strengths, as this allows us to provide very granular controls over actions. It also means that many
+of the decisions we make are very verbose. Take ssh access for example: an 'allow' decision for ssh access must include
+parameters for agent forwarding, port forwarding, concurrent connection limits, BPF events, expired cert disconnect, locking,
+file copying, etc. For this reason we will eschew any attempt at creating a "unified" method for decisions in favor of
+implementing custom methods and types intended to meet the specific needs of specific decisions.
+
+In order to ensure that the PDP API remains simple and usable we will establish a set of conventions for how decision methods
+and types should be handled:
+
+- All "allow" decisions will be structured as conditional allows, pending the application of limits/parameters. All allow decisions
+will be communicated by a type of the form `<Action>Permit`, which will encode all limits/parameters to be applied by the PEP. Much of
+the agent-side teleport logic that previously would have worked with some combination of certificate identity and role set will instead
+take the appropriate permit structure as input. Permits will *always* be passed by pointer, not value, as a precaution against zero
+value bugs resulting in unintended access.
```

Comment:
nit: the pointer/value distinction seems much lower-level than the rest of this RFD and kind of debatable, I wonder if it's better left out. If anything I wonder if you might want the fields within the Permit struct to be pointers instead of values?